### PR TITLE
json.dump of FITS_rec with VLF succeeds but produces unexpected result

### PR DESCRIPTION
**Pull request by volodymyrss**
### Description

This pull request is to address jsonification of `FITS_rec` containing variable length arrays, represented as [VLF class](https://github.com/astropy/astropy/blob/81c1d646da98730064a6dc86ccadb7bed9045cf2/astropy/io/fits/column.py#L1963). As it is inherited from `np.recarray`, it jsonifies with custom encoder included in astropy, [JsonCustomEncoder](https://docs.astropy.org/en/stable/api/astropy.utils.misc.JsonCustomEncoder.html).
Which relies on `np.ndarray.tolist`.

However, `VLF` enclosed in `FITS_rec` ends up represented as a different list (containing description of the variable length array - sizes and offsets).
This leads to json dump silently producing undesirable values.

I am not sure if it is anyway wise to rely in this jsonification, but I thought it at least it should not produce something unexpected without an error.

Please see the included test for details.

Looking forward for any feedback!

I am not sure if I should have created an issue first, if yes - sorry.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label.
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Review by pllim — COMMENTED** [review]
Thanks! Looks like that code hasn't really changed in the last 10 years since @embray transferred it from PyFITS... 👀 

Some general feedback thus far:

* Change log system has changed. Please see https://github.com/astropy/astropy/blob/main/docs/changes/README.rst . You also need to squash out your changes to `CHANGES.rst`.
* Please use `git rebase` instead of `git pull` to avoid "merge commits".

I'll hold off on approving CI run for now, but will be happy to when it is ready to run them.

**Comment by pllim** [comment]
Not sure about the milestone, I'll let @saimn decide.

**Comment by volodymyrss** [comment]
Thanks for the feedback! 

* I moved the changelog (still keeping two identical entries in different categories, not sure if it's ok).
* I rebased

**Comment by pllim** [comment]
@volodymyrss , can you also please squash all six commits into one? We do not want the history of modiyfing CHANGES.rst to exist. Thank you for your patience! 🙏

**Comment by volodymyrss** [comment]
> @volodymyrss , can you also please squash all six commits into one? We do not want the history of modiyfing CHANGES.rst to exist. Thank you for your patience! pray

Sure, @pllim!

**Comment by volodymyrss** [comment]
Sorry @pllim, @saimn, could I be so rude to remind of the status of this issue? I understood from the guide it is acceptable sometimes.
Thanks for all your work!

**Comment by pllim** [comment]
Ops... sorry it slipped through the cracks and thanks for the ping! I am not familiar with the FITS standards but I have pinged other maintainers to review this. Thanks for your patience!

File: astropy/io/fits/fitsrec.py
Comment (saimn) [inline astropy/io/fits/fitsrec.py:1296]: Could be shorter with a list comprehension:
```
column_lists = [column.array.tolist() for column in self.columns]
```

File: astropy/utils/tests/test_misc.py
Comment (saimn) [inline astropy/utils/tests/test_misc.py:108]: Could you test first that `data.tolist()` is returning the expected result ? And move the test to `io.fits`.

Comment (saimn) [reply astropy/utils/tests/test_misc.py:108]: Also the `i2` column should not have an object dtype.

File: docs/changes/utils/11957.bugfix.rst
Comment (saimn) [inline docs/changes/utils/11957.bugfix.rst:1]: Can you remove this one? The change concerns `io.fits`.

**Review by saimn — CHANGES_REQUESTED** [review]
Thanks @volodymyrss, and sorry I forgot about this one. I think the changes is really just `io.fits` because you're fixing `tolist` and the custom encoder is just one way to serialize that relies on this method. Also I think we can see this as a bug fix so I updated the milestone.

File: astropy/io/fits/fitsrec.py
Comment (volodymyrss) [reply astropy/io/fits/fitsrec.py]: Ok, that's how I did it at first but then unfolded it a way which some people might find more readable. No problem to keep it as comprehension.

File: astropy/utils/tests/test_misc.py
Comment (volodymyrss) [reply astropy/utils/tests/test_misc.py:108]: > Also the `i2` column should not have an object dtype.

good point, thanks.

Comment (volodymyrss) [reply astropy/utils/tests/test_misc.py:109]: > Could you test first that `data.tolist()` is returning the expected result ? And move the test to `io.fits`.

It looks like tolist() is not explicitly tested or even particularly used in `io.fits` yet. I am adding a test in there. There is an additional complexity which I did not intend to address at this time, since it is has no impact  on json, but I will address it now.

So I wanted to have this json serialization comparison test which is independent of particular implementation. 
Should I keep this json comparison? If yes, should I really move it to `io.fits`?  Since jsonification is tested in utils.

File: docs/changes/utils/11957.bugfix.rst
Comment (volodymyrss) [reply docs/changes/utils/11957.bugfix.rst:1]: Should there be a test of json serialization, and if yes, should it be in `io.fits`, an if yes, would this change (only of the test) warrant change in `utils`?

File: astropy/io/fits/tests/test_core.py
Comment (saimn) [inline astropy/io/fits/tests/test_core.py:1406]: This should go in `test_table:TestVLATables`.

File: astropy/utils/tests/test_misc.py
Comment (saimn) [reply astropy/utils/tests/test_misc.py:109]: `tolist` is part of the API of `.data` tables, so making sure it works also with VLA is a good thing to do.

Re- json test, yes I think it's useful. Both make sense actually, testing that fits objects can be serialized to json or that the `JsonCustomEncoder` can serialize fits objects. But currently there is no test with fits in `test_misc.py` and the VLA is a quite specific feature. It would actually be useful to have a more general test for `JsonCustomEncoder` and fits, with different types of columns. Either location is fine to me.

Comment (volodymyrss) [reply astropy/utils/tests/test_misc.py]: Ok, I started to make a test with some column variants [here](https://github.com/oda-hub/astropy/blob/ef1b4bf71/astropy/utils/tests/test_misc.py#L89-L113). Please let me know if it does not fit your intent, otherwise I will continue later.

File: docs/changes/utils/11957.bugfix.rst
Comment (volodymyrss) [reply docs/changes/utils/11957.bugfix.rst]: Ok, I made some set of tests.

File: astropy/utils/tests/test_misc.py
Comment (saimn) [inline astropy/utils/tests/test_misc.py:126]: This seems a bit complicated. My suggestion would be to use some test files we already have, e.g. these two ones:
```
❯ showtable astropy/io/fits/tests/data/btable.fits 
order    name     mag   Sp 
----- ---------- ----- ----
    1     Sirius -1.45  A1V
    2    Canopus -0.73 F0Ib
    3 Rigil Kent  -0.1  G2V

❯ showtable astropy/io/fits/tests/data/variable_length_table.fits 
   var     xyz [2]
---------- -------
   [45 56] 11 .. 3
[11 12 13] 12 .. 4
```
Just reading them and comparing the serialized string to a reference one.

**Review by saimn — CHANGES_REQUESTED** [review]
Thanks for working on the tests @volodymyrss, a few more comments below.

File: astropy/io/fits/column.py
Comment (volodymyrss) [reply astropy/io/fits/column.py:2014]: Ok! Should E201 be added to code style then, eventually? Like, in a separate PR?

Comment (pllim) [reply astropy/io/fits/column.py]: If you hook up `flake8` to your IDE, the check is pretty strict, which would inform you of this:

https://github.com/astropy/astropy/blob/a9411363fa07346eaa821a842f8c8c645af9036d/setup.cfg#L152-L154

It is more laxed in CI for historical reasons:

https://github.com/astropy/astropy/blob/a9411363fa07346eaa821a842f8c8c645af9036d/tox.ini#L161

Comment (volodymyrss) [reply astropy/io/fits/column.py]: thanks @pllim, yes, I have flake8 in my ide, but with even looser settings (for unrelated and possibly questionable reasons).

I did `tox -e codestyle` before committing. And it did not complain, as well as the CI. For the reason you mention, of course.
Which is why I meant to ask if maybe it would be useful for me to make an PR adding E201? 
Sorry, I do not know how style-only changes are handled here. Maybe it's too off-topic anyway.

Comment (pllim) [reply astropy/io/fits/column.py]: > make an PR adding E201

You are free to submit a PR, but you might also open a can of worms in case there are failures from code that was added like 10 years ago. 😬 

> how style-only changes are handled here

We currently handled it, uh, yolo-style? There is no style, but depends on who is looking at what, their mood, and how attentive they are to the diff. 😉

Comment (pllim) [reply astropy/io/fits/column.py]: Ah, but maybe adding E201 is much less controversial. I forgot we have that now.

https://github.com/astropy/astropy/blob/main/.pep8speaks.yml

Comment (volodymyrss) [reply astropy/io/fits/column.py]: > > make an PR adding E201
> 
> You are free to submit a PR, but you might also open a can of worms in case there are failures from code that was added like 10 years ago. grimacing

I tried, and it's only 16 problems in total given by `tox -e codestyle`. I will make a PR, a bit later.

> 
> > how style-only changes are handled here
> 
> We currently handled it, uh, yolo-style? There is no style, but depends on who is looking at what, their mood, and how attentive they are to the diff. wink

Comment (pllim) [reply astropy/io/fits/column.py]: > primary style guidelines

Once in a while, we fantasize about making the tox check stricter but we don't want to deal with all the code that has been there many years. You brought up a good point about making this guideline clearer somewhere. Maybe such expectations could be made clearer over at https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions -- @embray ?

Comment (volodymyrss) [reply astropy/io/fits/column.py]: > > primary style guidelines
> 
> Once in a while, we fantasize about making the tox check stricter but we don't want to deal with all the code that has been there many years. You brought up a good point about making this guideline clearer somewhere. Maybe such expectations could be made clearer over at https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions -- @embray ?

Thanks!

It says in there on the link to do `tox -e codestyle`, that's where I learned this is how to enforce guidelines. If it's ok, I will make the PR with adding E201 (and E202) and those style 16 fixes. I can deal with it later https://github.com/astropy/astropy/pull/12009 .

File: astropy/utils/tests/test_misc.py
Comment (saimn) [reply astropy/utils/tests/test_misc.py:126]: Since this is tested with `tolist` in `io.fits`, no need to repeat the duplicate test here.

Comment (volodymyrss) [reply astropy/utils/tests/test_misc.py:126]: > Since this is tested with `tolist` in `io.fits`, no need to repeat the duplicate test here.

I made two tests for `tolist` in `io.fits` [here](https://github.com/oda-hub/astropy/blob/7edebe69b32a0b3b2491f271d0e79b9975926e23/astropy/io/fits/tests/test_table.py#L2954-L2968).

But as I understood from discussion [above](https://github.com/astropy/astropy/pull/11957#discussion_r680117162) it may be useful to have an test of jsonification independent of particular implementation.
Or should there tests for both file-read and constructed cases with tolist but only file for json?

Comment (saimn) [reply astropy/utils/tests/test_misc.py:126]: For json testing only with files is fine. What we want to test here is that different types of columns are correctly serialized, it doesn't matter how the table is constructed since this is tested in the `io.fits` test you added.

Comment (volodymyrss) [reply astropy/utils/tests/test_misc.py:126]: I made the tests for json dumping of files here https://github.com/astropy/astropy/pull/11957/commits/c2b36a2577d0e501df1270bf7301d73b90ff9ed7

If you insist, I remove the `test_JsonCustomEncoder_FITS_rec_columns` [there](https://github.com/oda-hub/astropy/blob/c2b36a2577d0e501df1270bf7301d73b90ff9ed7/astropy/utils/tests/test_misc.py#L90), no problem.

Comment (volodymyrss) [reply astropy/utils/tests/test_misc.py:126]: > For json testing only with files is fine. What we want to test here is that different types of columns are correctly serialized, it doesn't matter how the table is constructed since this is tested in the `io.fits` test you added.

Sorry did not see this before  my last comment.
It seems to me that the table is actually somehow different depending on whether it is read or constructed. What differs is how what is in the array. If `tolist` is now outputs the same - it does not matter. But, this aspect of implementation of the custom encoder may change in the future. I am curious to look a bit more what is the nature of this difference, but I could just delete the longer non-file test, no problem.

Comment (saimn) [reply astropy/utils/tests/test_misc.py]: With the latest implementation of `tolist` the behavior is the same, and the json serializer the `io.fits` internal doesn't matter. So thanks for having removed the other test, looks good now.

File: astropy/io/fits/fitsrec.py
Comment (saimn) [inline astropy/io/fits/fitsrec.py:1290]: Could you rephrase to something simpler, e.g. "Override .tolist to take care of VLF" and remove the TODO comment below ?

**Comment by saimn** [comment]
Also please squash the commits after addressing the comment above, thanks.

File: astropy/utils/tests/test_misc.py
Comment (volodymyrss) [reply astropy/utils/tests/test_misc.py]: It could happen in the future that json serializer will use something else than tolist, and the test will not break when expected.  Presumably, at that time a test will be added.

File: astropy/io/fits/fitsrec.py
Comment (volodymyrss) [reply astropy/io/fits/fitsrec.py:1290]: Without explaining what's special about VLF? Ok.

**Comment by volodymyrss** [comment]
> Also please squash the commits after addressing the comment above, thanks.

Ok, all squashed.

**Review by saimn — APPROVED** [review]
All good, thanks @volodymyrss !

**Comment by volodymyrss** [comment]
Thanks @pllim  @saimn !

**Pull request merged**
